code/crates/test/src/middleware.rs

```diff
@@ -62,3 +72,103 @@ pub trait Middleware: fmt::Debug + Send + Sync {
 pub struct DefaultMiddleware;
 
 impl Middleware for DefaultMiddleware {}
+
+#[derive(Copy, Clone, Debug)]
+pub struct RotateValidators {
+    pub selection_size: usize,
+}
+
+impl Middleware for RotateValidators {
+    // Selects N validators from index height % num_validators (circularly).
+    // Example:
+    //   - N = 3, num_validators = 5, height = 0 -> [0, 1, 2]
+    //   - N = 3, num_validators = 5, height = 3 -> [3, 4, 0]
+    fn get_validator_set(
+        &self,
+        _ctx: &TestContext,
+        _current_height: Height,
+        height: Height,
+        genesis: &Genesis,
+    ) -> Option<ValidatorSet> {
+        let num_validators = genesis.validator_set.len();
+
+        if num_validators <= self.selection_size {
+            return Some(genesis.validator_set.clone());
+        }
+
+        let start_index = height.as_u64() as usize % num_validators;
+        let end_index = start_index + self.selection_size;
+
+        let selected = genesis
+            .validator_set
+            .get_all()
+            .into_iter()
+            .cycle()
+            .skip(start_index)
+            .take(end_index - start_index)
+            .collect::<Vec<_>>();
```

Comment:
This can be rewritten as 

```rust
        let start_index = height.as_u64() as usize % num_validators;

        let selected = genesis
            .validator_set
            .iter()
            .cycle()
            .skip(start_index)
            .take(self.selection_size)
            .cloned()
            .collect::<Vec<_>>();
```

if you replace `get_all` by this:

```rust
    /// Iterate over the validators in the set
    pub fn iter(&self) -> slice::Iter<Validator> {
        self.validators.iter()
    }
```

Reply:
This can maybe be extracted into a standalone function to share its code between both middleware